### PR TITLE
DateBox doesn't save typed value if it's opened (T833467)

### PR DESCRIPTION
**Pull request by ksercs**

### CODE REVIEW

File: js/ui/date_box/ui.date_box.strategy.calendar.js [hunk 1]
```diff
@@ -147,6 +151,7 @@ var CalendarStrategy = DateBoxStrategy.inherit({
         if(dateBox.option("applyValueMode") === "instantly") {
             dateBox.option("opened", false);
             this.dateBoxValue(this.getValue(), e.event);
+            this._lastActionElement = null;
```
Comment (DokaRus) [inline js/ui/date_box/ui.date_box.strategy.calendar.js]: Should we reset the `_lastActionElement` when dropDown closed?

Comment (ksercs) [reply js/ui/date_box/ui.date_box.strategy.calendar.js]: You are right, it's not necessary cause this variable will be set to "calendar" or "input" after any action while dropDown is opened

File: js/ui/date_box/ui.date_box.strategy.calendar.js [hunk 2]
```diff
@@ -28,7 +28,8 @@ var CalendarStrategy = DateBoxStrategy.inherit({
 
                     if(this._widget.option("zoomLevel") === this._widget.option("maxZoomLevel")) {
                         var contouredDate = this._widget._view.option("contouredDate");
-                        contouredDate && this.dateBoxValue(contouredDate, e);
+                        var lastActionElement = this._lastActionElement;
+                        contouredDate && lastActionElement === "calendar" && this.dateBoxValue(contouredDate, e);
```
Comment (San4es) [inline js/ui/date_box/ui.date_box.strategy.calendar.js]:
```suggestion
                        if(contouredDate && lastActionElement === "calendar") {
                            this.dateBoxValue(contouredDate, e);
                        }
```

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 3]
```diff
@@ -379,6 +379,44 @@ QUnit.module("datebox tests", moduleConfig, () => {
         assert.equal($input.val(), "", "text is not rendered");
     });
 
+    QUnit.test(`After typing while calendar is opened the typed data should be saved`, assert => {
+        var optionsSet = [];
```
Comment (DokaRus) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]:
```suggestion
        const optionsSet = [];
```

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 4]
```diff
@@ -379,6 +379,44 @@ QUnit.module("datebox tests", moduleConfig, () => {
...
+                optionsSet.push({
+                    useMaskBehavior,
+                    type,
+                    pickerType: "calendar",
+                    penOnFieldClick: true
+                });
+            });
+        });
+
+        optionsSet.forEach(function(options) {
```
Comment (DokaRus) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]:
```suggestion
        optionsSet.forEach((options) => {
```

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 5]
```diff
@@ -379,6 +379,44 @@ QUnit.module("datebox tests", moduleConfig, () => {
...
+            const $dateBox = $("#dateBox").dxDateBox(
+                options
+            );
+
+            const instance = $dateBox.dxDateBox("instance");
+            const $input = $dateBox.find("." + TEXTEDITOR_INPUT_CLASS);
```
Comment (DokaRus) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]:
```suggestion
            const $input = $dateBox.find(`.${TEXTEDITOR_INPUT_CLASS}`);
```

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 6]
```diff
@@ -379,6 +379,44 @@ QUnit.module("datebox tests", moduleConfig, () => {
...
+            kb.type(typedDate).press("enter");
+            assert.deepEqual(instance.option("text"), typedDate, `typed value is set when useMaskBehavior:${options.useMaskBehavior}, type:${options.type}`);
+
+            instance.open();
+            kb
+                .keyDown('left', { ctrlKey: true })
+                .press('right')
+                .press('enter');
+            assert.deepEqual(instance.option("text"), selectedDate, `value is successfully changed by calendar when useMaskBehavior:${options.useMaskBehavior}, type:${options.type}`);
+        }.bind(this));
```
Comment (DokaRus) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]:
```suggestion
        });
```

**Review by DokaRus — APPROVED** [review]

**Review by San4es — APPROVED** [review]

**Pull request merged**